src/integrationtests/java/com/aws/greengrass/integrationtests/ShadowManagerTest.java

```diff
@@ -247,4 +264,51 @@ void GIVEN_shadow_manager_WHEN_strategy_config_resets_THEN_respond_to_config_upd
         kernel.getContext().waitForPublishQueueToClear();
         assertThat(syncHandler.getOverallSyncStrategy(), instanceOf(PeriodicSyncStrategy.class));
     }
+
+    @Test
+    void GIVEN_cryptoKeyProviderService_WHEN_get_client_THEN_wait_for_cryptoservice(ExtensionContext context) throws InterruptedException, TLSAuthException, IoTDataPlaneClientCreationException {
+        ignoreExceptionOfType(context, TLSAuthException.class);
+        ignoreExceptionOfType(context, RetryableException.class);
+        ignoreExceptionOfType(context, ResourceNotFoundException.class);
+
+        IotDataPlaneClientFactory factory = spy(kernel.getContext().get(IotDataPlaneClientFactory.class));
+        kernel.getContext().put(IotDataPlaneClientFactory.class, factory);
+        IotDataPlaneClientWrapper wrapper = spy(new FakeIotDataPlaneClientWrapper(factory));
+        kernel.getContext().put(IotDataPlaneClientWrapper.class, wrapper);
+
+        startNucleusWithConfig(NucleusLaunchUtilsConfig.builder()
+                .configFile("sync.yaml")
+                .mqttConnected(true)
+                .mockCloud(false)
+                .build());
+        BaseSyncStrategy s = kernel.getContext().get(RealTimeSyncStrategy.class);
+        assertThat("syncing has started", s::isSyncing, eventuallyEval(is(true)));
+        verify(wrapper, timeout(5000).atLeast(1)).getThingShadow("Thing1", "");
+
+        CountDownLatch cdl=new CountDownLatch(1);
+        SecurityService ss= mock(SecurityService.class);
+        when(ss.getDeviceIdentityKeyManagers()).thenAnswer((invocation)->{
+            cdl.countDown();
+            return new KeyManager[0];
+        });
+        kernel.getContext().put(SecurityService.class, ss);
+        assertThat("request is retried with a new client",cdl.await(10, TimeUnit.SECONDS), is(true));
+        verify(wrapper, timeout(5000).atLeast(2)).getThingShadow("Thing1", "");
+    }
+
+    static class FakeIotDataPlaneClientWrapper extends  IotDataPlaneClientWrapper{
+        IotDataPlaneClientFactory factory;
+
+        public FakeIotDataPlaneClientWrapper(IotDataPlaneClientFactory iotDataPlaneClientFactory) {
+            super(iotDataPlaneClientFactory);
+            factory = iotDataPlaneClientFactory;
+        }
+
+        @Override
+        public GetThingShadowResponse getThingShadow(String thingName, String shadowName) throws IoTDataPlaneClientCreationException {
```

Comment:
We should go one level deeper instead of creating a fake at this level. Doing this will hide away potential failures on the Wrapper coed. We create and `IotDataPlaneClient` fake that throws the error you expect that way we are only faking the boundaries

Reply:
Agreed!

Reply:
However, in this case, it gets tricky as creation of a new `IotDataPlaneClient` is part of the code in test itself and moreover that is an interface. It is not injected as such.

Reply:
Yeah I see. We could still create a fake on the factory and only override the getClientMethod